File: tools/bench/window_max_deque.cpp
// II-5 — 슬라이딩 윈도우 최댓값: 창을 매번 훑는 방법과 덱으로 후보만 남기는 방법 (C++ 쪽).
//
// window_max_deque.py 와 1:1 로 대응한다. 같은 n, 같은 k, 같은 난수 규칙.
//
// 보려는 것은 절대 시간이 아니라 k 에 대한 기울기다. 창마다 다시 훑으면
// O(nk) 라 k 가 4배가 되면 시간도 약 4배가 되고, 덱은 O(n) 이라 k 를 키워도
// 시간이 거의 그대로다.
//
// 빌드/실행:
//   g++ -std=c++17 -O2 tools/bench/window_max_deque.cpp -o /tmp/wm && /tmp/wm

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <deque>
#include <random>
#include <vector>
using namespace std;
using clk = chrono::steady_clock;

static const int N = 200000;

static vector<int> naive(const vector<int>& a, int k) {
    vector<int> out;
    out.reserve(a.size() - k + 1);
    for (size_t i = 0; i + k <= a.size(); ++i)
        out.push_back(*max_element(a.begin() + i, a.begin() + i + k));
    return out;
}

static vector<int> with_deque(const vector<int>& a, int k) {
    deque<int> dq;
    vector<int> out;
    out.reserve(a.size() - k + 1);
    for (size_t i = 0; i < a.size(); ++i) {
        while (!dq.empty() && a[dq.back()] <= a[i]) dq.pop_back();
        dq.push_back((int)i);
        if (dq.front() <= (int)i - k) dq.pop_front();
        if ((int)i >= k - 1) out.push_back(a[dq.front()]);
    }
    return out;
}

template <typename F>
static double bench(F fn, const vector<int>& a, int k, int repeat = 3) {
    vector<double> ts;
    for (int r = 0; r < repeat; ++r) {
        auto t0 = clk::now();
        auto v = fn(a, k);
        auto t1 = clk::now();
        if (v.empty()) puts("");  // 최적화로 통째로 사라지는 것을 막는다
        ts.push_back(chrono::duration<double>(t1 - t0).count());
    }
    sort(ts.begin(), ts.end());
    return ts[ts.size() / 2];
}

int main() {
    mt19937 rng(20250804);
    vector<int> a(N);
    for (int& x : a) x = (int)(rng() % 1000000000u) + 1;

    printf("n = %d  (g++ 13 -O2)\n", N);
    printf("%8s | %14s | %10s | %8s\n", "k", "창마다 훑기", "덱", "배수");
    puts("------------------------------------------------");
    for (int k : {100, 400, 1600}) {
        double t1 = bench(naive, a, k);
        double t2 = bench(with_deque, a, k);
        printf("%8d | %11.4f초 | %8.4f초 | %6.1f배\n", k, t1, t2, t1 / t2);
    }
    return 0;
}

File: tools/bench/window_max_deque.py
#!/usr/bin/env python3
"""II-5 — 슬라이딩 윈도우 최댓값: 창을 매번 훑는 방법과 덱으로 후보만 남기는 방법.

두 방법은 같은 답을 낸다. 다른 것은 창이 커질 때의 기울기다. 창마다 다시
훑으면 O(nk) 라 k 에 비례해 느려지고, 덱은 각 원소가 한 번 들어가고 한 번
나오므로 O(n) 이라 k 와 무관하다. 표의 "k 를 4배로 늘렸을 때" 열이 그 증거다.

`window_max_deque.cpp` 와 1:1 로 대응한다.

사용법:
    python3.13 tools/bench/window_max_deque.py
"""
from __future__ import annotations

import random
import time
from collections import deque

N = 200_000


def naive(a: list[int], k: int) -> list[int]:
    """창마다 다시 훑는다. max(슬라이스)는 C 루프라 순수 파이썬보다 훨씬 빠르지만
    그래도 원소를 k번 본다는 사실은 바뀌지 않는다."""
    return [max(a[i : i + k]) for i in range(len(a) - k + 1)]


def with_deque(a: list[int], k: int) -> list[int]:
    """덱에는 '아직 최댓값이 될 수 있는 후보'의 인덱스만 남긴다."""
    dq: deque[int] = deque()
    out = []
    for i, x in enumerate(a):
        while dq and a[dq[-1]] <= x:
            dq.pop()
        dq.append(i)
        if dq[0] <= i - k:
            dq.popleft()
        if i >= k - 1:
            out.append(a[dq[0]])
    return out


def bench(fn, a: list[int], k: int, repeat: int = 3) -> float:
    ts = []
    for _ in range(repeat):
        t = time.perf_counter()
        fn(a, k)
        ts.append(time.perf_counter() - t)
    return sorted(ts)[len(ts) // 2]


def main() -> None:
    random.seed(20250804)
    a = [random.randrange(1, 10**9) for _ in range(N)]

    assert naive(a[:5000], 100) == with_deque(a[:5000], 100)

    print(f"n = {N:,}  (Python 3.13)")
    print(f"{'k':>8} | {'창마다 훑기':>12} | {'덱':>10} | {'배수':>8}")
    print("-" * 48)
    for k in (100, 400, 1600):
        t1 = bench(naive, a, k)
        t2 = bench(with_deque, a, k)
        print(f"{k:>8,} | {t1:>11.3f}초 | {t2:>9.3f}초 | {t1 / t2:>7.1f}배")


if __name__ == "__main__":
    main()
